server/src/main/java/org/apache/druid/server/coordinator/DruidCoordinatorRuntimeParams.java

```diff
@@ -346,22 +352,35 @@ public Builder withDataSources(Collection<ImmutableDruidDataSource> dataSourcesC
       return this;
     }
 
+    /** This method must be used in test code only. */
     @VisibleForTesting
-    public Builder withAvailableSegments(DataSegment... availableSegments)
+    public Builder withAvailableSegmentsInTest(DataSegment... availableSegments)
     {
-      this.availableSegments.addAll(Arrays.asList(availableSegments));
-      return this;
+      return withAvailableSegmentsInTest(Arrays.asList(availableSegments));
     }
 
-    public Builder withAvailableSegments(Collection<DataSegment> availableSegments)
+    /** This method must be used in test code only. */
+    @VisibleForTesting
+    public Builder withAvailableSegmentsInTest(Collection<DataSegment> availableSegments)
     {
-      this.availableSegments.addAll(Collections.unmodifiableCollection(availableSegments));
+      TreeSet<DataSegment> availableSegmentSet = createAvailableSegmentsSet();
+      availableSegmentSet.addAll(availableSegments);
+      return setAvailableSegments(availableSegmentSet);
+    }
+
+    public Builder setAvailableSegments(TreeSet<DataSegment> availableSegments)
```

Comment:
Semantics of `withAvailableSegments()` are different from `setAvailableSegments()`: the former makes a (defensive) copy of the passed collection. The latter doesn't. Those who call `setAvailableSegments()` should be aware that if they modify the collection that they passed into this method afterwards, the "internal" set of segments for Params will be updated too (simply because this is the same object).

I'll add a Javadoc comment.

Reply:
Thanks for explanation and javadoc, makes sense :+1: